### DISCUSSION TITLE
Attributes instead of interfaces

### DISCUSSION THREAD

**Discussion by roxblnfk** [container-interop/service-provider#69] [Ideas]
Hello. It's nice to see the designing of PSR in a rational way.

Have you discussed marking factories with attributes, instead of declaring interfaces that return closures? In this case, even Symfony with their compiled container could squeeze the maximum out of it.

```php
class ServiceProvider {
    /**
     * Define container ID
     */
    #[Factory(id: 'alias')]
    public function makeAService(Psr\Container\ContainerInterface $container): mixed {...}

    /**
     * Without ID the return type may be used
     */
    #[Factory]
    public function makeFooService(Psr\Container\ContainerInterface $container): FooService {...}

    /**
     * Despite your insistence that the closure should only take the container, 
     * people will still want autowiring. This is also great for compiled containers.
     */
    #[Factory]
    public function makeWithAutowiring(App\BarService $dependency): FooService {...}
}
```

Attributes can also be extensible. This way different implementations can add their own settings, but it will still be compatible with PSR. I haven't thought about how correct this is yet. I would most likely use a second attribute with some specific settings alongside.

**Comment by roxblnfk** [comment]
Ah, found it in early discussions.
https://github.com/container-interop/service-provider/discussions/64#discussioncomment-7929713

However, note that this is the path that might not meet serious resistance from Symfony.

**Reply by roxblnfk** [reply]
Hi @mindplay-dk 👋
I've noticed that you've decided to revisit this concept. That's great.

**Reply by mindplay-dk** [reply]
Yes. 😄

Having come to the realization that there are essentially two model based approaches: a provider model is either declarative (AST/DSL style approaches) or functional (callables) - you might could picture hybrid approaches between those two, but these would essentially be functional too; the difference, in a nutshell, is whether the models include behavior (functions) or data-only.

The idea here is to move away from providers per se, towards pure factory classes - in a sense, this approach is "just PHP code" with some metadata, it doesn't favorize either callable-based or compiled containers, instead it just kind of lays out factory classes and says "you figure it out".

In a sense, it's equally suitable/unsuitable for both types of containers, but it's more "neutral territory" in the sense that it's "just code", just factories, and doesn't assume or resemble anything that happens inside containers.

I've had to open my mind a bit 😅

I also explored the idea of a pure AST/DSL style model - you can read [here](https://groups.google.com/g/php-fig/c/SWIfYEkw89I/m/CHxR8BYGAAAJ) why that idea won't work.

We know the callable-based approach won't work for the compiled containers due to performance constraints.

So that leaves this idea as the last option that I'm aware of. Seeing as how you can take this more complex format and adapt it to the existing proposed interfaces, I initially dismissed this idea as being largely an opinionated version of what we already had. But it seems like this might be the only approach that could potentially satisfy both the compiler/performance constraint and could still interop with extensions in callable-based containers.

Of course, for compiled containers, this hinges on using an actual AST-based code parser, essentially recompiling the code - this should be easier to do with flat methods though, than it would be with callables. And compiled containers need to compile anyway - their own provider formats typically have to resemble an AST in order to be able to compile anyway, so why not use the AST of the language itself.

There is a long way to go for such a proposal though. Many details and constraints to be worked out.

My current thinking is to proof-of-concept this by defining the attributes, writing some factory classes, using them in some callable-based containers (the easy part) and then building a library that extracts a generic AST-like provider model from factory classes, exposing a provider-like model that includes class-names, source-code expressions, etc... data structures suitable for bootstrapping compiled containers. (the hard part.)

But it's not likely I'll go ahead with this work unless there is some indication of interest from the major frameworks, and then by the FIG - some FIG core members have expressed interest, but conditional on interest from the major frameworks... so we'll see. 🙂

**Reply by roxblnfk** [reply]
> My current thinking is to proof-of-concept this by defining the attributes, writing some factory classes, using them in some callable-based containers (the easy part) and then building a library that extracts a generic AST-like provider model from factory classes, exposing a provider-like model that includes class-names, source-code expressions, etc... data structures suitable for bootstrapping compiled containers. (the hard part.)

I think that compiling the correct set of attributes is the most challenging task in the service providers designing.
Do you already have a draft list of attributes and their semantic descriptions?

> But it's not likely I'll go ahead with this work unless there is some indication of interest from the major frameworks, and then by the FIG - some FIG core members have expressed interest, but conditional on interest from the major frameworks... so we'll see. 🙂

Major frameworks themselves do not know what they need, so engaging them will not be easy. I am confident that you will find greater cooperation from the Yii3 side, at least I am pushing them towards it. So, if you manage to sway Symfony, everything will work out.

**Reply by mindplay-dk** [reply]
> I am confident that you will find greater cooperation from the Yii3 side, at least I am pushing them towards it.

Alexander Makarov (samdark) has already expressed interest and participated in discussions. 🙂

> So, if you manage to sway Symfony, everything will work out.

Symfony and Laravel will be much harder to sway.

I think, if we went ahead and built everything, including adapters for the major frameworks, and assuming the community showed interest, and packages started to ship with factory classes -- then they would have to at least address it, wouldn't they. 😏

But even if I built it all, I don't have the social media reach to make it happen. And it's a massive undertaking - I'm not sure I have the time or the energy to do it all myself.

I would love to be part of a working group though, so that's my hope. It's difficult to engage the right people to make that happen though. I'm just a guy with no name who happens to build stuff. 😄

**Reply by mindplay-dk** [reply]
> I think that compiling the correct set of attributes is the most challenging task in the service providers designing.
Do you already have a draft list of attributes and their semantic descriptions?

sorry, missed that one 😉

no, the idea is quite new, I haven't spent much time on it yet - all I have is this [rough draft](https://gist.github.com/mindplay-dk/cd5bf33533a2b7c6fe96469a1b9bd8f6) but that was literally just 20 minutes of doodling, by no means a real design...

**Reply by roxblnfk** [reply]
Yes, I've seen that draft 👌
I think the `Extend` and `Inject` attributes are superfluous there, and it would be better to focus on what relates to factories and service providers.
Perhaps I am mistaken, but currently `Extend` and `Inject` look like answering "how to inject" instead of "how to produce"

**Reply by mindplay-dk** [reply]
@roxblnfk I don't know what you mean? When the component isn't a singleton (resolved by the type-hint alone) the container needs some way of knowing which service ID to inject - and for extensions, it additionally needs to know which service ID you're expecting to extend and replace. How would it work without these?

**Reply by roxblnfk** [reply]
@mindplay-dk Indeed 👌 When I was thinking about this, I fell out of the context of service providers and imagined that users would use these attributes in the code of their services 😃
Perhaps, it would be a good idea to limit the scope of attribute usage right in the RFC.

**Comment by mindplay-dk** [comment]
Thank you for your participation. 🙂

As you've discovered, we discussed this idea already - it is a cool idea, and I have some experimental code in a repository playing around with some ideas in this area myself.

I would certainly welcome experiments in this area - a service provider builder implementing patterns like these, I'm sure, would be a very cool proof of concept, and a very nice thing to have (and show) if this PSR comes to fruition! 👍

**Reply by roxblnfk** [reply]
I think we'll implement it in [Spiral](https://github.com/spiral/framework)
